Do not send to channel when close in coreapi

We try to send data to a closed  channel. We are missing the fuse box in front of the send process.

```
panic: send on closed channel 
goroutine 229 [running]: 
code.vegaprotocol.io/vega/core/coreapi/services.(*NetParams).Push(0xc001829600, {0xc002eff4f0?, 0x1, 0x0?}) 
        /home/runner/work/vega/vega/core/coreapi/services/netparams.go:74 +0x125 
code.vegaprotocol.io/vega/core/broker.(*Broker).startSending.func1(0xc00481f4a0, 0x1a) 
        /home/runner/work/vega/vega/core/broker/broker.go:227 +0x5fb 
created by code.vegaprotocol.io/vega/core/broker.(*Broker).startSending 
        /home/runner/work/vega/vega/core/broker/broker.go:201 +0x5cf
```

## Solution

We should:
1. have a wait.group waiting in front of the close([a.ch](http://a.ch/))
2. verification of the closed state, in front of the consume() function

## Testing

According to @wwestgarth , it's liekly this code path is hit when we have an unknown netparams in the genesis file. That should help with testing.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -77,6 +77,8 @@
 - [10215](https://github.com/vegaprotocol/vega/issues/10215) - Rework pagination to align with the natural reverse-chronological order of the block explorer.
 - [10241](https://github.com/vegaprotocol/vega/issues/10241) - Do not include start epoch on referees set statistics.
 - [10219](https://github.com/vegaprotocol/vega/issues/10219) - Candles API should fill gaps when there are periods with no trades.
+- [10050](https://github.com/vegaprotocol/vega/issues/10050) - Cleanup `mempool` cache on commit. 
+- [9882](https://github.com/vegaprotocol/vega/issues/9882) - Fix `net params` sent on closed channel.
 
 ## 0.73.0
 

--- a/core/coreapi/services/netparams.go
+++ b/core/coreapi/services/netparams.go
@@ -35,6 +35,7 @@ type NetParams struct {
 
 	mu        sync.RWMutex
 	netParams map[string]vegapb.NetworkParameter
+	isClosed  bool
 	ch        chan vegapb.NetworkParameter
 }
 
@@ -49,13 +50,32 @@ func NewNetParams(ctx context.Context) (netParams *NetParams) {
 }
 
 func (a *NetParams) consume() {
-	defer func() { close(a.ch) }()
+	defer func() {
+		// clear the channel before we close it
+		// can't use a WaitGroup as suggested because if we get to this point
+		// if the Push is still in progress and the queue is full, we'll
+		// end up in a situation where we're waiting for the queue to empty, but
+		// no consumer.
+		for {
+			select {
+			case _, ok := <-a.ch:
+				if !ok {
+					close(a.ch)
+					return
+				}
+			}
+		}
+	}()
+
 	for {
 		select {
 		case <-a.Closed():
+			a.mu.Lock()
+			a.isClosed = true
+			a.mu.Unlock()
 			return
 		case netParams, ok := <-a.ch:
-			if !ok {
+			if !ok || a.isClosed {
 				// cleanup base
 				a.Halt()
 				// channel is closed
@@ -71,6 +91,9 @@ func (a *NetParams) consume() {
 func (a *NetParams) Push(evts ...events.Event) {
 	for _, e := range evts {
 		if ae, ok := e.(netParamsE); ok {
+			if a.isClosed {
+				return
+			}
 			a.ch <- ae.NetworkParameter()
 		}
 	}

--- a/core/coreapi/services/netparams.go
+++ b/core/coreapi/services/netparams.go
@@ -56,15 +56,11 @@ func (a *NetParams) consume() {
 		// if the Push is still in progress and the queue is full, we'll
 		// end up in a situation where we're waiting for the queue to empty, but
 		// no consumer.
-		for {
-			select {
-			case _, ok := <-a.ch:
-				if !ok {
-					close(a.ch)
-					return
-				}
-			}
+		for range a.ch {
+			// do nothing just let the sender finish what it's doing and empty the channel
 		}
+
+		close(a.ch)
 	}()
 
 	for {

--- a/core/coreapi/services/netparams.go
+++ b/core/coreapi/services/netparams.go
@@ -91,9 +91,12 @@ func (a *NetParams) consume() {
 func (a *NetParams) Push(evts ...events.Event) {
 	for _, e := range evts {
 		if ae, ok := e.(netParamsE); ok {
+			a.mu.Lock()
 			if a.isClosed {
+				a.mu.Unlock()
 				return
 			}
+			a.mu.Unlock()
 			a.ch <- ae.NetworkParameter()
 		}
 	}

--- a/core/coreapi/services/netparams_test.go
+++ b/core/coreapi/services/netparams_test.go
@@ -6,12 +6,11 @@ import (
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
-
-	"code.vegaprotocol.io/vega/core/events"
-	"github.com/stretchr/testify/require"
-
 	"code.vegaprotocol.io/vega/core/coreapi/services"
+	"code.vegaprotocol.io/vega/core/events"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestNetParams(t *testing.T) {

--- a/core/coreapi/services/netparams_test.go
+++ b/core/coreapi/services/netparams_test.go
@@ -1,3 +1,18 @@
+// Copyright (C) 2023 Gobalsky Labs Limited
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
 package services_test
 
 import (

--- a/core/coreapi/services/netparams_test.go
+++ b/core/coreapi/services/netparams_test.go
@@ -1,0 +1,50 @@
+package services_test
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"code.vegaprotocol.io/vega/core/events"
+	"github.com/stretchr/testify/require"
+
+	"code.vegaprotocol.io/vega/core/coreapi/services"
+)
+
+func TestNetParams(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	np := services.NewNetParams(ctx)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	allSent := false
+
+	maxEvents := 1000000
+
+	evts := make([]events.Event, maxEvents)
+
+	for i := 0; i < maxEvents; i++ {
+		evts[i] = events.NewNetworkParameterEvent(ctx, "foo", "bar")
+	}
+
+	require.NotPanics(t, func() {
+		go func() {
+			np.Push(
+				evts...,
+			)
+			allSent = true
+			wg.Done()
+		}()
+	})
+
+	// slight pause to give the goroutine a chance to start pushing before we cancel the context
+	time.Sleep(time.Millisecond)
+	cancel()
+
+	wg.Wait()
+
+	assert.True(t, allSent)
+}

